// xgplayer/es/plugins/download/saveAs.js
import { objectSpread2 as _objectSpread2, slicedToArray as _slicedToArray, typeof as _typeof } from "../../_virtual/_rollupPluginBabelHelpers.js";
import sniffer from "../../utils/sniffer.js";
var DEFAULT_CONFIG = {
  mimeType: "application/octet-stream",
  fileName: "download",
  timeout: 3e4
};
var BrowserUtils = {
  supportsBlob: function supportsBlob() {
    return typeof Blob !== "undefined";
  },
  supportsObjectURL: function supportsObjectURL() {
    return typeof URL !== "undefined" && URL.createObjectURL;
  },
  isSafari: function isSafari() {
    return sniffer.browser === "safari";
  },
  isIEWithMsSaveBlob: function isIEWithMsSaveBlob() {
    return navigator.msSaveBlob;
  },
  supportsDownloadAttribute: function supportsDownloadAttribute() {
    var anchor = document.createElement("a");
    return "download" in anchor;
  }
};
var DataURLUtils = {
  isDataURL: function isDataURL(str) {
    return /^data:([\w+-]+\/[\w+.-]+)?[,;]/.test(str);
  },
  dataURLToBlob: function dataURLToBlob(dataURL) {
    var parts = dataURL.split(/[:;,]/);
    var type = parts[1];
    var indexDecoder = dataURL.indexOf("charset") > 0 ? 3 : 2;
    var decoder = parts[indexDecoder] === "base64" ? atob : decodeURIComponent;
    var binData = decoder(parts.pop());
    var uiArr = new Uint8Array(binData.length);
    for (var i = 0; i < binData.length; i++) {
      uiArr[i] = binData.charCodeAt(i);
    }
    var BlobConstructor = window.Blob || window.MozBlob || window.WebKitBlob || function(parts2, opts) {
      return parts2;
    };
    return new BlobConstructor([uiArr], {
      type
    });
  },
  hasNonASCII: function hasNonASCII(str) {
    return /([\x80-\xff])/.test(str);
  },
  stringToUint8Array: function stringToUint8Array(str) {
    var arr = new Uint8Array(str.length);
    for (var i = 0; i < str.length; i++) {
      arr[i] = str.charCodeAt(i);
    }
    return arr;
  }
};
var DownloadUtils = {
  downloadWithAnchor: function downloadWithAnchor(url, fileName) {
    var anchor = document.createElement("a");
    anchor.href = url;
    anchor.setAttribute("download", fileName);
    anchor.className = "download-js-link";
    anchor.innerHTML = "downloading...";
    anchor.style.display = "none";
    var clickHandler = function clickHandler2(e) {
      e.stopPropagation();
      this.removeEventListener("click", clickHandler2);
    };
    anchor.addEventListener("click", clickHandler);
    document.body.appendChild(anchor);
    setTimeout(function() {
      anchor.click();
      document.body.removeChild(anchor);
      if (url.startsWith("blob:")) {
        setTimeout(function() {
          return URL.revokeObjectURL(url);
        }, 250);
      }
    }, 66);
    return true;
  },
  downloadWithSafari: function downloadWithSafari(url, fileName) {
    if (url instanceof Blob) {
      var reader = new FileReader();
      reader.onload = function(e) {
        var dataURL = e.target.result;
        DownloadUtils.downloadWithSafari(dataURL, fileName);
      };
      reader.readAsDataURL(url);
      return true;
    }
    if (DataURLUtils.isDataURL(url)) {
      url = "data:" + url.replace(/^data:([\w\/\-\+]+)/, DEFAULT_CONFIG.mimeType);
    }
    if (!window.open(url)) {
      var confirmed = confirm("Displaying New Document\n\nUse Save As... to download, then click back to return to this page.");
      if (confirmed) {
        location.href = url;
      }
    }
    return true;
  },
  downloadWithIframe: function downloadWithIframe(url, fileName) {
    if (url instanceof Blob) {
      var reader = new FileReader();
      reader.onload = function(e) {
        var dataURL = e.target.result;
        DownloadUtils.downloadWithIframe(dataURL, fileName);
      };
      reader.readAsDataURL(url);
      return true;
    }
    var iframe = document.createElement("iframe");
    document.body.appendChild(iframe);
    if (DataURLUtils.isDataURL(url)) {
      url = "data:" + url.replace(/^data:([\w\/\-\+]+)/, DEFAULT_CONFIG.mimeType);
    }
    iframe.src = url;
    setTimeout(function() {
      return document.body.removeChild(iframe);
    }, 333);
    return true;
  }
};
var RequestUtils = {
  createRequest: function createRequest(url) {
    var options = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
    var xhr = new XMLHttpRequest();
    if (options.timeout) {
      xhr.timeout = options.timeout;
    }
    if (typeof options.withCredentials === "boolean") {
      xhr.withCredentials = options.withCredentials;
    }
    return xhr;
  },
  setHeaders: function setHeaders(xhr) {
    var headers = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
    Object.entries(headers).forEach(function(_ref) {
      var _ref2 = _slicedToArray(_ref, 2), key = _ref2[0], value = _ref2[1];
      xhr.setRequestHeader(key, value);
    });
  },
  downloadFromURL: function downloadFromURL(url, fileName) {
    var options = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {};
    return new Promise(function(resolve, reject) {
      var xhr = RequestUtils.createRequest(url, options);
      xhr.open("GET", url, true);
      xhr.responseType = "blob";
      if (options.headers) {
        RequestUtils.setHeaders(xhr, options.headers);
      }
      if (options.withCredentials) {
        xhr.withCredentials = true;
      }
      xhr.onload = function(e) {
        if (e.target.status === 200) {
          saveAs(e.target.response, fileName, options.mimeType || DEFAULT_CONFIG.mimeType);
          resolve(e.target.response);
        } else {
          reject(new Error("HTTP ".concat(e.target.status, ": ").concat(e.target.statusText)));
        }
      };
      xhr.onerror = function() {
        reject(new Error("Network error occurred"));
      };
      xhr.ontimeout = function() {
        reject(new Error("Request timeout"));
      };
      if (options.beforeRequest) {
        options.beforeRequest(xhr);
      }
      setTimeout(function() {
        return xhr.send();
      }, 0);
    });
  }
};
function saveAs(data, fileNameOrOptions, mimeType) {
  var options = {};
  var fileName = DEFAULT_CONFIG.fileName;
  var actualMimeType = DEFAULT_CONFIG.mimeType;
  if (typeof fileNameOrOptions === "string") {
    fileName = fileNameOrOptions || DEFAULT_CONFIG.fileName;
    actualMimeType = mimeType || DEFAULT_CONFIG.mimeType;
  } else if (_typeof(fileNameOrOptions) === "object" && fileNameOrOptions !== null) {
    options = _objectSpread2(_objectSpread2({}, DEFAULT_CONFIG), fileNameOrOptions);
    fileName = options.fileName || DEFAULT_CONFIG.fileName;
    actualMimeType = options.mimeType || DEFAULT_CONFIG.mimeType;
  }
  if (String(this) === "true") {
    var temp = data;
    data = fileNameOrOptions;
    fileNameOrOptions = temp;
    actualMimeType = mimeType || DEFAULT_CONFIG.mimeType;
  }
  if (typeof data === "string" && !fileNameOrOptions && !mimeType && data.length < 2048) {
    var url = data;
    var extractedFileName = url.split("/").pop().split("?")[0];
    fileName = extractedFileName || DEFAULT_CONFIG.fileName;
    var anchor = document.createElement("a");
    anchor.href = url;
    if (anchor.href.indexOf(url) !== -1) {
      return RequestUtils.downloadFromURL(url, fileName, options);
    }
  }
  if (DataURLUtils.isDataURL(data)) {
    if (data.length > 1024 * 1024 * 1.999 && BrowserUtils.supportsBlob()) {
      var _blob = DataURLUtils.dataURLToBlob(data);
      actualMimeType = _blob.type || actualMimeType;
      return saveAs(_blob, fileName, actualMimeType);
    } else {
      return BrowserUtils.isIEWithMsSaveBlob() ? navigator.msSaveBlob(DataURLUtils.dataURLToBlob(data), fileName) : DownloadUtils.downloadWithAnchor(data, fileName);
    }
  }
  if (typeof data === "string" && DataURLUtils.hasNonASCII(data)) {
    var uint8Array = DataURLUtils.stringToUint8Array(data);
    var BlobConstructor = window.Blob || window.MozBlob || window.WebKitBlob || function(parts, opts) {
      return parts;
    };
    data = new BlobConstructor([uint8Array], {
      type: actualMimeType
    });
  }
  var blob;
  if (data instanceof Blob) {
    blob = data;
  } else {
    var _BlobConstructor = window.Blob || window.MozBlob || window.WebKitBlob || function(parts, opts) {
      return parts;
    };
    blob = new _BlobConstructor([data], {
      type: actualMimeType
    });
  }
  if (BrowserUtils.isIEWithMsSaveBlob()) {
    return navigator.msSaveBlob(blob, fileName);
  }
  if (BrowserUtils.supportsObjectURL()) {
    var objectURL = URL.createObjectURL(blob);
    return DownloadUtils.downloadWithAnchor(objectURL, fileName);
  }
  if (BrowserUtils.isSafari()) {
    return DownloadUtils.downloadWithSafari(blob, fileName);
  }
  if (typeof blob === "string" || blob.constructor === String) {
    try {
      var dataURL = "data:" + actualMimeType + ";base64," + btoa(blob);
      return DownloadUtils.downloadWithAnchor(dataURL, fileName);
    } catch (e) {
      var _dataURL = "data:" + actualMimeType + "," + encodeURIComponent(blob);
      return DownloadUtils.downloadWithAnchor(_dataURL, fileName);
    }
  }
  var reader = new FileReader();
  reader.onload = function(e) {
    DownloadUtils.downloadWithAnchor(e.target.result, fileName);
  };
  reader.readAsDataURL(blob);
  return DownloadUtils.downloadWithIframe(blob, fileName);
}
function saveAsWithAuth(url) {
  var options = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
  var config = _objectSpread2(_objectSpread2({}, DEFAULT_CONFIG), options);
  var fileName = config.fileName || url.split("/").pop().split("?")[0] || "download";
  return RequestUtils.downloadFromURL(url, fileName, config);
}
export { saveAs, saveAsWithAuth };
